### PR TITLE
DEV-2546 Add smoke tests for each input mode

### PR DESCRIPTION
**Pull request by pauldwolfe**
Creates new smoke tests to cover all 3 input modes and their expected files. To 
facilitate this a couple of changes to the smoke test framework itself:
- Remove API interaction and use json to simplify setup
- Remove assertion on cram contents (might revisit this later)
- Add a simple runner to parallelize tests
- Create test dir for each input type with input and expected output. 
- Add some magic to enable local runs without changing code. 

While testing this also found some issues with the implementation of modes, so those
changes are in here too.

### CODE REVIEW

File: cluster/src/main/java/com/hartwig/pipeline/metadata/InputMode.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package com.hartwig.pipeline.metadata;
 
 public enum InputMode {
-    TUMOR_NORMAL,
+    TUMOR_REFERENCE,
```
[the remark this answers is not in the document]
Comment (nedleitch) [reply cluster/src/main/java/com/hartwig/pipeline/metadata/InputMode.java:4]: Maybe `TUMOR_AND_REFERENCE` would allay my concerns at the price of some brevity.

File: cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java [hunk 2]
```diff
@@ -59,48 +58,67 @@ public void tearDown() throws Exception {
...
-                .build();
+                .outputBucket("smoketest-pipeline-output-pilot-1")
+                .context(Context.DIAGNOSTIC);
+
+        final String username = System.getProperty("user.name");
+        if (username.equals("root")) {
+            String privateKeyPath = workingDir() + "/google-key.json";
+            builder.privateKeyPath(privateKeyPath).uploadPrivateKeyPath(privateKeyPath);
+        } else {
+            builder.cloudSdkPath(String.format("/Users/%s/google-cloud-sdk/bin", username));
```
Comment (nedleitch) [inline cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java:95]: The Hartwig Linux Users Group may be stirred to contribute a patch for this.

File: cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java [hunk 3]
```diff
@@ -59,48 +58,67 @@ public void tearDown() throws Exception {
...
         cleanupBucket(setName, arguments.outputBucket(), storage);
     }
 
+    @NotNull
+    private String version(final String inputMode) {
+        String version = System.getProperty("version");
+        if (version.equals("local-SNAPSHOT")) {
+            version = System.getProperty("user.name") + "-" + inputMode;
+        }
+        return version.length() > 14 ? version.substring(0, 14) : version;
```
Comment (nedleitch) [inline cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java:119]: This is a bit cryptic, I assume the `14` refers to the length of the string `local-SNAPSHOT`?

**Review by nedleitch — CHANGES_REQUESTED** [review]

File: cluster/src/main/java/com/hartwig/pipeline/calling/sage/SageGermlineCaller.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+public class SageGermlineCaller extends SageCaller {
+
+    public SageGermlineCaller(final AlignmentPair alignmentPair, final PersistedDataset persistedDataset,
+            final ResourceFiles resourceFiles) {
+        super(alignmentPair, persistedDataset, SageConfiguration.germline(resourceFiles));
+    }
+
+    @Override
+    public List<BashCommand> tumorOnlyCommands(final SomaticRunMetadata metadata) {
+        return Collections.emptyList();
```
[the remark this answers is not in the document]
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/calling/sage/SageGermlineCaller.java:21]: Actually it should not be implemented, we won't run sage germline in tumor only. I can remove it.

File: cluster/src/main/java/com/hartwig/pipeline/execution/vm/VirtualMachineJobDefinition.java [hunk 5]
```diff
@@ -79,24 +79,16 @@ static VirtualMachineJobDefinition germlineCalling(BashStartupScript startupScri
...
     }
 
-    static VirtualMachineJobDefinition sageSomaticCalling(BashStartupScript startupScript, ResultsDirectory resultsDirectory) {
+    static VirtualMachineJobDefinition sageCalling(final BashStartupScript startupScript, final ResultsDirectory resultsDirectory,
+            final String namespace) {
         return ImmutableVirtualMachineJobDefinition.builder()
-                .name("sage-somatic")
+                .name(namespace.replace("_", "-"))
                 .startupCommand(startupScript)
                 .performanceProfile(custom(32, 120))
```
[the remark this answers is not in the document]
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/execution/vm/VirtualMachineJobDefinition.java:87]: Good point, i had thought they were the same.

File: cluster/src/main/java/com/hartwig/pipeline/metadata/InputMode.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pauldwolfe) [reply cluster/src/main/java/com/hartwig/pipeline/metadata/InputMode.java:4]: I also prefer tumor_normal, but reference is everywhere in hmftools and the api so its really hard to make it consistent that way. I could add the AND.

File: cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java (hunk 3 above)
Comment (pauldwolfe) [reply cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java:119]: Yes, i'll extract a constant.

File: cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java (hunk 2 above)
Comment (pauldwolfe) [reply cluster/src/test/java/com/hartwig/pipeline/smoke/SmokeTest.java:95]: apologies. i can only think of one possible solution https://www.apple.com/ca/macbook-pro/

**Pull request merged**